src/libos/crates/host-socket/src/datagram/receiver.rs

```diff
@@ -68,20 +93,57 @@ impl<A: Addr, R: Runtime> Receiver<A, R> {
         if A::domain() == Domain::Netlink {
             let recv_bytes = inner.try_copy_buf_netlink(bufs);
             if let Some(recv_bytes) = recv_bytes {
-                let recv_addr = inner.get_addr().unwrap();
+                let recv_addr = inner.get_packet_addr();
                 if inner.recv_len.is_none() {
                     // All recv_len are consumed, do next host recv.
                     self.do_recv(&mut inner);
                 }
 
-                return Ok((recv_bytes, recv_addr));
+                return Ok((recv_bytes, recv_addr, 0, 0));
             }
         } else {
-            let recv_bytes = inner.try_copy_buf(bufs);
-            if let Some(recv_bytes) = recv_bytes {
-                let recv_addr = inner.get_addr().unwrap();
-                self.do_recv(&mut inner);
-                return Ok((recv_bytes, recv_addr));
+            let copied_bytes = inner.try_copy_buf(bufs);
+            if let Some(copied_bytes) = copied_bytes {
+                let recv_addr = inner.get_packet_addr();
+                // Copy ancillary data from control buffer
+                let msg_controllen = inner.req.msg.msg_controllen;
+                if msg_controllen > super::OPTMEM_MAX {
+                    return_errno!(EINVAL, "invalid msg control length");
+                }
+                let control_buf_len = if control.is_some() {
+                    control.as_ref().unwrap().len()
+                } else {
+                    0
+                };
+                assert!(msg_controllen >= control_buf_len);
+
+                if msg_controllen > 0 {
+                    control
```

Comment:
The `MSG_CTRUNC` flag should be set when some control data is discarded.

Reply:
Done